File: BlockOfLabs3/laboratory3.3/lab3.3.3.cpp
#include <iostream>
#include <string>
#include <fstream>
using namespace std;
const int FILE_KEY = 1;
const int MIN_ARR_SIZE = 1;
const int CONSOLE_KEY = 2;
const int MIN_FILE_WAY_SIZE = 5;



/// text blocks
void conditionOutput()
{
	cout << "The program is designed to sort an array\n\tusing the simple insertion method.\n\n";
}
void pathConditionOutput()
{
	cout << "Where will we work through:\n\tFile: " << FILE_KEY << " Console: " << CONSOLE_KEY << endl << endl;
}
void fileRestriction()
{
	cout << "\n*The first number is the number of elements\nof the array, and subsequent numbers of this array*\n";
}



/// is can open
bool isCanOpenFile(string way, ios_base::openmode mode)
{
	fstream file(way, mode);
	file.close();
	return file.good();
}



/// input way of work
int choosingAPath()
{
	int path = 0;
	bool isIncorrect = true;

	pathConditionOutput();
	cout << "Please write were we should work: ";
	do
	{
		cin >> path;

		if (cin.fail() || cin.get() != '\n')
		{
			cerr << "Error. You should write a one natural number. Try again: ";
			cin.clear();
			while (cin.get() != '\n');
		}
		else
		{
			if (path == CONSOLE_KEY || path == FILE_KEY)
				isIncorrect = false;
			else cerr << "Error method. Try again: ";
		}
	} while (isIncorrect);

	return path;
}



/// input way to the file
bool pathCondition(string way)
{
	if (way.size() < MIN_FILE_WAY_SIZE)
	{
		cerr << "The path is too short. Try again: ";
		return false;
	}
	string bufstr = way.substr(way.size() - 4);
	if (bufstr != ".txt")
	{
		cerr << "Write .txt file. Try again: ";
		return false;
	}
	return true;
}
string inputPath()
{
	string way;
	bool isIncorrect;
	do
	{
		cin >> way;
		isIncorrect = !pathCondition(way);
	} while (isIncorrect);
	return way;
}
string inputPathToTheFile()
{
	string fileWay;
	bool isIncorrect = true;
	do
	{
		bool isCorrect = true;
		fileWay = inputPath();
		if (!isCanOpenFile(fileWay, ios::in))
			cout << "Can't open a file. Try write another way: ";
		else
			isIncorrect = false;
	} while (isIncorrect);
	return fileWay;
}



/// input from console
int arrSizeInputFromConsole()
{
	int arrSize;
	bool isIncorrect = true;

	cout << "Write your arr size: ";
	do
	{
		cin >> arrSize;
		if (cin.fail() || cin.get() != '\n')
		{
			cout << "Invalid numeric input. Try again: ";
			cin.clear();
			while (cin.get() != '\n');
		}
		else if (arrSize < MIN_ARR_SIZE)
			cout << "Minimal arr size is: " << MIN_ARR_SIZE <<
			". Try again: ";
		else
			isIncorrect = false;
	} while (isIncorrect);

	return arrSize;
}
int inputCurrentNumbFromConsole()
{
	int currentNumb;
	bool isIncorrect = true;
	do
	{
		cin >> currentNumb;
		if (cin.fail() || cin.get() != '\n')
		{
			cout << "Invalid numeric input. Try again: ";
			cin.clear();
			while (cin.get() != '\n');
		}
		else
			isIncorrect = false;
	} while (isIncorrect);

	return currentNumb;
}
bool isCorrectArrOfNumbInputFromConsole(int arrSize, int*& arrOfNumb)
{
	for (int i = 0; i < arrSize; i++)
	{
		cout << "Write your " << i + 1 << " number: ";
		arrOfNumb[i] = inputCurrentNumbFromConsole();
	}

	return false;
}



/// input from file
int arrSizeInputFromFile(string fileWay)
{
	int arrSize;
	ifstream file(fileWay, ios::in);
	file >> arrSize;

	if (file.fail() || file.get() != '\n')
	{
		cout << "Error in input size of massive. Try again: ";
		arrSize = -1;
	}
	else if (arrSize < MIN_ARR_SIZE)
	{
		cout << "Minimal arr size is " << MIN_ARR_SIZE << ". Try again: ";
		arrSize = -1;
	}

	file.close();

	return arrSize;
}
bool isIncorrectArrOfNumbInputFromFile(int arrSize, int*& arrOfNumb, string fileWay)
{
	if (arrSize == -1)
		return true;

	ifstream file(fileWay, ios::in);
	for (int i = 0; i < arrSize; i++)
	{
		file >> arrOfNumb[i];
		if (file.fail())
		{
			file.close();
			cout << "Error in reading massive elements. Try again: ";
			return true;
		}
	}
	file.close();


	return false;
}



void sortMassive(int*& arrOfNumb, int arrSize)
{
	int temp; // временная переменная для хранения значения элемента сортируемого массива

	for (int i = 1; i < arrSize; i++)
	{
		temp = arrOfNumb[i]; // инициализируем временную переменную текущим значением элемента массива

		int j = i - 1; // запоминаем индекс предыдущего элемента массива
		while (j >= 0 && arrOfNumb[j] > temp) // пока индекс не равен 0 и предыдущий элемент массива больше текущего
		{
			arrOfNumb[j + 1] = arrOfNumb[j]; // перестановка элементов массива
			arrOfNumb[j] = temp;
			j--;
		}
	}
}



/// output block
void outputFromFile(int* arrOfNumb, int arrSize)
{
	bool isIncorrect = true;
	cout << "Write way to your file: ";
	do
	{
		string fileWay = inputPath();
		if (isCanOpenFile(fileWay, ios::out))
		{
			ofstream file(fileWay, ios::out);
			for (int i = 0; i < arrSize; i++)
				file << arrOfNumb[i] << " ";
			cout << "Check your file.";
			file.close();
			isIncorrect = false;
		}
		else
			cerr << "Can't open a file. Try write another way: ";
	} while (isIncorrect);
}
void outputFromConsole(int* arrOfNumb, int arrSize)
{
	for (int i = 0; i < arrSize; i++)
		cout << arrOfNumb[i] << " ";
}
void resultOutput(int* arrOfNumb, int arrSize)
{
	cout << "You need to choose where to write information from.\n";
	int path = choosingAPath();

	path == FILE_KEY ? outputFromFile(arrOfNumb, arrSize) : outputFromConsole(arrOfNumb, arrSize);
}



int arrSizeInput(int path, string fileWay) {
	return path == FILE_KEY ? arrSizeInputFromFile(fileWay) : arrSizeInputFromConsole();
}

bool inputArrOfNumb(int arrSize, int*& arrOfNumb, int path, string fileWay) {
	return path == FILE_KEY ? isIncorrectArrOfNumbInputFromFile(arrSize, arrOfNumb, fileWay) : isCorrectArrOfNumbInputFromConsole(arrSize, arrOfNumb);
}

int main() {
	int arrSize;
	int* arrOfNumb;

	conditionOutput();

	cout << "\nYou need to choose where to read information from.\n";
	int path = choosingAPath();

	bool isIncorrect;
	do {
		string fileWay = path == FILE_KEY ? inputPathToTheFile() : "";

		arrSize = arrSizeInput(path, fileWay);
		arrOfNumb = new int[arrSize];
		isIncorrect = inputArrOfNumb(arrSize, arrOfNumb, path, fileWay);
	} while (isIncorrect);

	sortMassive(arrOfNumb, arrSize);

	resultOutput(arrOfNumb, arrSize);

	delete[] arrOfNumb;
	arrOfNumb = nullptr;

	return 0;
}

File: BlockOfLabs3/laboratory3.3/lab3.3.3.java
package lab3;

import java.io.File;
import java.io.FileWriter;
import java.util.Scanner;

public class Lab3 {
    static final int FILE_KEY = 1;
    static final int CONSOLE_KEY = 2;
    static final  int MIN_ARR_SIZE = 1;
    static final int MIN_FILE_WAY_SIZE = 5;

    static void conditionOutput()
    {
        System.out.println("""
        The program is designed to sort an array
                using the simple insertion method.
        """);
    }

    static void pathConditionOutput()
    {
        System.out.printf("""
                Where will we work through:\s
                        File: %d Console: %d
                        
                """, FILE_KEY, CONSOLE_KEY);
    }

    static void fileRestriction()
    {
        System.out.print("""
                        
                *The first number is the number of elements
                of the array, and subsequent numbers of this array*
                Write way to your file:\s""");
    }

    static int choosingAPath(Scanner in)
    {
        int path = 0;
        boolean isIncorrect = true;

        pathConditionOutput();

        System.out.print("Please write were we should work: ");
        do
        {
            boolean isCorrect = false;

            try
            {
                path = Integer.parseInt(in.nextLine());
                isCorrect = true;
            } catch(Exception error)
            {
                System.err.print("Error. You should write a one natural number. Try again: ");
            }

            if (path == CONSOLE_KEY || path == FILE_KEY)
            {
                isIncorrect = false;
            }
            else if (isCorrect)
            {
                System.err.print("Error method. Try again: ");
            }
        } while (isIncorrect);

        return path;
    }

    static int arrSizeInputFromConsole(Scanner in)
    {
        int arrSize = 0;
        boolean isIncorrect = true;

        System.out.print("Write your arr size: ");
        do
        {
            boolean isCorrect = false;

            try
            {
                arrSize = Integer.parseInt(in.nextLine());
                isCorrect = true;
            } catch(Exception error)
            {
                System.err.print("Invalid numeric input. Try again: ");
            }

            if (arrSize < MIN_ARR_SIZE && isCorrect)
            {
                System.err.printf("Minimal arr size is: %d. Try again: ", MIN_ARR_SIZE);
            }
            else if (isCorrect)
            {
                isIncorrect = false;
            }
        } while (isIncorrect);

        return arrSize;
    }

    static int inputCurrentNumbFromConsole(Scanner in)
    {
        int currentNumb = 0;
        boolean isIncorrect = true;

        do
        {
            try
            {
                currentNumb = Integer.parseInt(in.nextLine());
                isIncorrect = false;
            } catch(Exception error)
            {
                System.err.print("Invalid numeric input. Try again: ");
            }

        } while (isIncorrect);

        return currentNumb;
    }

    static boolean isCorrectArrOfNumbInputFromConsole(int arrSize, int[] arrOfNumb, Scanner in)
    {
        for (int i = 0; i < arrSize; i++)
        {
            System.out.printf("Write your %d number: ", i + 1);
            arrOfNumb[i] = inputCurrentNumbFromConsole(in);
        }

        return false;
    }

    static boolean isCanOpenFile(String way)
    {
        File file = new File(way);

        return file.canRead();
    }

    static boolean pathCondition(String way)
    {
        if (way.length() < MIN_FILE_WAY_SIZE)
        {
            System.err.print("The path is too short. Try again: ");
            return false;
        }

        String bufstr = way.substring(way.length() - 4);
        if (!bufstr.equals(".txt"))
        {
            System.err.print("Write .txt file. Try again: ");
            return false;
        }

        return true;
    }

    static String inputPath(Scanner in)
    {
        String way;
        boolean isIncorrect;

        do
        {
            way = in.nextLine();
            isIncorrect = !pathCondition(way);
        } while (isIncorrect);

        return way;
    }


    static String inputPathToTheFile(Scanner in)
    {
        String fileWay;
        boolean isIncorrect = true;

        do
        {
            fileWay = inputPath(in);

            if (!isCanOpenFile(fileWay))
            {
                System.err.print("Can't open a file. Try write another way: ");
            }
            else
            {
                isIncorrect = false;
            }
        } while (isIncorrect);

        return fileWay;
    }

    static int arrSizeInputFromFile(String fileWay)
    {
        int arrSize;
        try (Scanner fileScanner = new Scanner(new File(fileWay))){
            arrSize = Integer.parseInt(fileScanner.nextLine());
        } catch(Exception error){
            System.err.print("Error in input size of massive. Try again: ");
            arrSize = -1;
        }
        if (arrSize < MIN_ARR_SIZE)
        {
            System.err.printf("Minimal arr size is %d. Try again: ", MIN_ARR_SIZE);
            arrSize = -1;
        }

        return arrSize;
    }

    static boolean isCorrectArrOfNumbInputFromFile(int arrSize,int[] arrOfNumb,String fileWay)
    {
        if (arrSize == -1)
            return true;

        try (Scanner fileScanner = new Scanner(new File(fileWay)))
        {
            fileScanner.nextInt();
            for (int i = 0; i < arrSize; i++)
            {
                arrOfNumb[i] = fileScanner.nextInt();
            }

            if (fileScanner.hasNext())
                return true;

        } catch(Exception error)
        {
            System.err.print("Error in reading massive elements. Try again: ");
            return true;
        }

        return false;
    }

    static void sortMassive(int[] arrOfNumb, int arrSize)
    {
        int temp;

        for (int i = 1; i < arrSize; i++)
        {
            temp = arrOfNumb[i];

            int j = i - 1;
            while (j >= 0 && arrOfNumb[j] > temp)
            {
                arrOfNumb[j + 1] = arrOfNumb[j];
                arrOfNumb[j] = temp;
                j--;
            }
        }
    }

    static void outputFromFile(int[] arrOfNumb, int arrSize, Scanner in)
    {
        boolean isIncorrect = true;

        System.out.print("Write way to your file: ");
        do
        {
            String fileWay = inputPath(in);
            File file = new File(fileWay);
            StringBuilder builder;
            if (file.canWrite())
            {
                try
                {
                    FileWriter writer = new FileWriter(fileWay);
                    builder = new StringBuilder();
                    for (int i = 0 ; i < arrSize; i++)
                        builder.append(arrOfNumb[i]).append(" ");
                    writer.write(builder.toString());
                    writer.close();
                    System.out.print("Check your file.");
                    isIncorrect = false;
                } catch (Exception error)
                {
                    System.err.print("Can't write in this file. Try again: ");
                }
            }
            else
            {
                System.err.print("Can't open a file. Try write another way: ");
            }
        } while (isIncorrect);
    }

    static void outputFromConsole(int[] arrOfNumb, int arrSize)
    {
        for (int i = 0; i < arrSize; i++)
        {
            System.out.printf("%d ", arrOfNumb[i]);
        }
    }

    static void resultOutput(int[] arrOfNumb, int arrSize, Scanner in)
    {
        System.out.println("You need to choose where to write information from.");
        int path = choosingAPath(in);
        if (path == CONSOLE_KEY)
        {
            outputFromConsole(arrOfNumb, arrSize);
        }
        else
        {
            outputFromFile(arrOfNumb, arrSize, in);
        }
    }

    static String inputFileWay(Scanner in, int path){
        return path == FILE_KEY ? inputPathToTheFile(in) : "";
    }

    static int arrSizeInput(Scanner in, int path, String fileWay){
        return path == FILE_KEY ? arrSizeInputFromFile(fileWay) : arrSizeInputFromConsole(in);
    }

    static boolean inputArrOfNumb(int arrSize,int[] arrOfNumb,int path,String fileWay,Scanner in){
        return path == FILE_KEY ? isCorrectArrOfNumbInputFromFile(arrSize, arrOfNumb, fileWay) : isCorrectArrOfNumbInputFromConsole(arrSize, arrOfNumb, in);
    }

    public static void main(String[] args)
    {
        Scanner in = new Scanner(System.in);

        int arrSize;
        int[] arrOfNumb = new int[0];

        conditionOutput();

        System.out.println("\nYou need to choose where to read information from.");

        int path = choosingAPath(in);

        boolean isIncorrect;
        if (path == FILE_KEY)
            fileRestriction();

        do {
            String fileWay = inputFileWay(in, path);
            arrSize = arrSizeInput(in, path, fileWay);
            if (arrSize != -1)
                arrOfNumb = new int [arrSize];
            isIncorrect = inputArrOfNumb(arrSize, arrOfNumb, path, fileWay, in);
        } while (isIncorrect);

        sortMassive(arrOfNumb, arrSize);

        resultOutput(arrOfNumb, arrSize, in);

        in.close();
    }
}